src/app/track/track.component.ts

```diff
@@ -1,346 +1,476 @@
-    const e = document.createEvent('MouseEvents');
-    // Use of deprecated function to satisfy TypeScript.
-    e.initMouseEvent('click', true, false,
-      document.defaultView ?? window, 0, 0, 0, 0, 0,
-      false, false, false, false, 0, null);
-    a.dispatchEvent(e);
-    a.remove();
-    this.isExporting = false;
-  }
-  populateComments() {
-    this.commentsService.getAll(this.track.slug)
-      .subscribe(comments => this.comments = comments);
-  }
-
-  populateTrackData() {
-    // this.trackData = this.trackDataService.getTrackData(this.track.slug);
-    this.trackDataService.getAll(this.track.slug)
-      .subscribe(trackData => { this.trackData = trackData; this.createMap(); });
-  }
-
-  addComment() {
-    this.isSubmitting = true;
-    this.commentFormErrors = {};
-
-    const commentBody = this.commentControl.value;
-    this.commentsService
-      .add(this.track.slug, commentBody)
-      .subscribe(
-        comment => {
-          this.comments.unshift(comment);
-          this.commentControl.reset('');
-          this.isSubmitting = false;
-        },
-        errors => {
-          this.isSubmitting = false;
-          this.commentFormErrors = errors;
+
+
+    exportGPX() {
+        this.isExporting = true;
+        let fileContents = '<?xml version="1.0"?>\n' +
+            '<gpx version="1.1" creator="openbikesensor.hlrs.de">\n';
+        fileContents += '<trk>\n';
+        fileContents += '<trkseg>\n';
+        let nameCounter = 1;
+        for (const p of this.trackData.points) {
+            fileContents += '<trkpt  lat="' + p.latitude + '" lon="' + p.longitude + '">\n' +
+                '<name>' + nameCounter + '</name>\n';
+            if (p.flag != null && p.flag > 0) {
+                fileContents += '<sym>Flag</sym>\n';
+            }
+            fileContents += '<cmt> d1="' + p.d1 + '" d2="' + p.d2 + '" </cmt>\n';
+            fileContents += '</trkpt>\n';
+            nameCounter++;
         }
-      );
-  }
-
-  onDeleteComment(comment) {
-    this.commentsService.destroy(comment.id, this.track.slug)
-      .subscribe(
-        success => {
-          this.comments = this.comments.filter((item) => item !== comment);
+        fileContents += '</trkseg>\n';
+        fileContents += '</trk>\n';
+        fileContents += '</gpx>';
+        const filename = 'track.gpx';
+        const filetype = 'text/plain';
+        const a = document.createElement('a');
+        const dataURI = 'data:' + filetype +
+            ';base64,' + btoa(fileContents);
+        a.href = dataURI;
+        a['download'] = filename;
+        const e = document.createEvent('MouseEvents');
+        // Use of deprecated function to satisfy TypeScript.
+        e.initMouseEvent('click', true, false,
+            document.defaultView ?? window, 0, 0, 0, 0, 0,
+            false, false, false, false, 0, null);
+        a.dispatchEvent(e);
+        a.remove();
+        this.isExporting = false;
+    }
+
+    exportCSV() {
```

Comment:
As I expect the exports to change more often than the rest of the component I think we should move them into another file as functions and only call the functions here to adhere more closely to the [Single Responsibility Principle](https://de.wikipedia.org/wiki/Single-Responsibility-Prinzip).
Ideally one file with a function for the OPX, and one for the CSV

Reply:
I have no clue how to do that in Angular as I'm new to TypeScript and Angular. Anyone get help here?

Reply:
IMHO ideally the component only transforms data for display but contains no business logic. But as you already said the exports will probably move into the backend anyway, I don't think we need to be so strict here.

Oh, and for someone new to TS and Angular this PR is damn good!